tailscale/datasource_devices_test.go

```diff
@@ -0,0 +1,134 @@
+package tailscale_test
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	"github.com/tailscale/terraform-provider-tailscale/tailscale"
+)
+
+func TestAccTailscaleDevices(t *testing.T) {
+	resourceName := "data.tailscale_devices.all_devices"
+
+	// This is a string containing tailscale_device datasource configurations
+	devicesDataSources := &strings.Builder{}
+
+	toResourceComponent := func(str string) string {
+		return strings.ReplaceAll(str, " ", "_")
+	}
+
+	// First test the tailscale_devices datasource, which will give us a list of
+	// all device IDs.
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviderFactories(t),
+		Steps: []resource.TestStep{
+			{
+				Config: `data "tailscale_devices" "all_devices" {}`,
+				Check: func(s *terraform.State) error {
+					client := testAccProvider.Meta().(*tailscale.Clients).V2
+					devices, err := client.Devices().List(context.Background())
+					if err != nil {
+						return fmt.Errorf("unable to list devices: %s", err)
+					}
+
+					devicesByID := make(map[string]map[string]any)
+					for _, device := range devices {
+						m := tailscale.DeviceToMap(&device)
```

Comment:
nit: probably just personal opinion, but I'd prefer to copy-paste a small utility function into tests rather than make it exported purely for tests. Or just make this test part of the `tailscale` package rather than a dedicated `_test` package.

Reply:
> Or just make this test part of the tailscale package rather than a dedicated _test package.

I started to go down that road, but there are reusable functions in the tailscale_test package that we want to use too. I'll look into maybe moving this into an `internal` package or something in a follow-on PR.